combat-trainer.lic

```diff
@@ -4347,6 +4360,34 @@ class GameState
     @equipment_manager.stow_weapon(weapon_training[skill])
   end
 
+  def engage(stomp_check = false)
+    if @stomp_on_cooldown && !Flags['war-stomp-ready'] && stomp_check # If we're using stomp on cooldown but stomp isn't ready
+      return
+    elsif (@stomp_on_cooldown || @stomp_to_engage) && Flags['war-stomp-ready'] # if we're using stomp and stomp is ready
```

Comment:
```suggestion
if (@stomp_on_cooldown || @stomp_to_engage) && Flags['war-stomp-ready'] # if we're using stomp and stomp is ready
```

Reply:
You don't need that first IF at all.

Reply:
We do, I added it for instances where, while using stomp on cooldown, we just want to stomp, and we're not seeking engage behavior. So, without that first one, every time it checks our cooldown, which would be frequently, we'd also engage. I obviously can't have it just return if the cooldown isn't up, because sometimes we want to go here to engage normally, even if we can't stomp.

Reply:
This logic isn't making sense to me. This is the engage def. We enter here when we want to engage. If war stomp isn't ready, we entered the engage section because we want to engage... it doesn't make sense to me to enter here just to stomp if we don't want to engage.

Reply:
So, we could tease out stomp and pounce to their own private methods, and redirect based on guild/settings/flags, but the idea is to use this method TO engage. Stomp on cooldown is a feature that's kind of it's own bag, since it can be reset and it does a ton of dmg. 

Another option is to relocate all the conditions to the check:
```ruby
game_state.engage(true) unless (game_state.npcs.empty? || !@stomp_on_cooldown)
```
would become
```ruby
game_state.engage if @stomp_on_cooldown && game_state.npcs && Flags['war-stomp-ready']
```
That would remove the need to even visit the method until we're ready to stomp, have something to stomp, and are stomping as often as possible.

Reply:
Submitted that, which removed that first condition you didn't like.

Reply:
I reckon having pounce and stomp be their own defs/actions, outside of the engage section would actually make sense here. In both cases, i think the idea would be to use them on cooldown anyway. So they should be separate checks, independent of engaging.

Reply:
So, lets break this down. First, separating them into their own methods. Code-wise, we'd basically make new methods with the same conditionals checked from inside engage:
```
def engage
  stomp if able and ready
  pounce if able and ready
  otherwise engage
end

def stomp
  return unless ready and able
  stomp
  if stomp fails
    engage if want to engage
  or
    return doing nothing
end

def pounce
  pounce
  if pounce fails
    engage
  end
end
```
So pounce is calling engage, stomp is calling engage, and vice versa. I'm not entirely sure what we gain by making them their own methods.

Secondly, regarding use, these are engage tools. I'd argue, having used stomp for a bit, that stomp to engage should be the default, and likely the same with pounce. It is instant melee, which is awesome. I don't actually think this should be opt in, though i'm not opposed to opting out, so we should leave the check for the setting. I imagine pounce would be the same, instant melee for no cost on a cooldown. 

Finally, wrt to making them independent, or using them on CD by default, I'd argue isn't a great idea. For stomp, using it to engage is great, but it can trigger back to back and wipe out the room, and your fatigue. Best use is likely just engage by default. 

Pounce, if it's just a ~3min cd, is likewise poorly utilized just spamming it every 3 minutes. If it does damage, we could add an option to use it on cd, but if it's strictly an engage tool, I think it should be used only in cases where we are hoping to engage. It's going to look mighty silly if we pounce from melee because it's off cooldown and 3 seconds later have to walk towards the next mob.

Reply:
Honestly I just pounce on cooldown anyway. It's just a tool to train Scouting in my view.  I have it as a buff nonspell now every 2m45s or whatever my cooldown is. 

I can see using it as an engaging tool I guess, that's just a preference thing. Sounds like war stomp is a bit different mechanically, and I don't have a barb to comment there. 

Sounds like the logic can be a bit more nuanced then. Engage when we need to, pounce if we're engaging from range? What about pole? Pounce to engage always?

Reply:
The training aspect changes things for pounce. How *much* does it train? is every 3 mins required? or do you get a big fat mindstate bump for each use?  

If we're getting locked in like 5 minutes, it doesn'tmake sense to treat it as a training tool. If it trains as poorly as hunt does now (for scouting/instinct), then on cooldown makes a helluva lot more sense than as a tool for engaging. Can i make a baby ranger and get pounce pretty quickly?

Reply:
heh, made a level 1 ranger on premie:
```
Gained: Instinct(+10), Light Armor(+1)
```
If that holds, we're back to engage tool. We could put it up where stomp is now, though,to use if 
1. off cooldown
2. npcs in the room
3. instinct mindstate is <24

So that we're using it on cooldown if we need the experience, then as an engagement tool thereafter.